### DISCUSSION TITLE
Idea - import packages.config into website package script builder

### DISCUSSION THREAD

**Discussion by csharpforevermore** [chocolatey/home#281] [Ideas]
I adore Chocolatey (and chocolate) so try to get the most out of it.

I also love the website script builder feature! Having just discovered the excellent `choco export `command (when I thought I could not love you more!) I wanna copy+paste **packages.config** to the script builder. Is it possible? 

**So basically I see the website's package script builder tool and the choco export output, as the two ends of the workflow.  They should be able to talk to each other**

This way I could easily modify or add to the script for use on my site.  Maybe also it would be good to bring it in line with the packages.config schema by allowing users to download such a file so they can do a choco install pakages.config instead of the choco commands. 

Alternatively the choco export could support an option to export the actual choco CLI commands as well?

**Comment by gep13** [comment]
@csharpforevermore thanks for reaching out!  Happy to hear that you are enjoying using Chocolatey!

> @csharpforevermore said...
This way I could easily modify or add to the script for use on my site. 

I actually think that this is a good idea.  @pauby what are your thoughts here?  @st3phhays would this work?

> @csharpforevermore said...
so they can do a choco install pakages.config instead of the choco commands.

Just to be clear, as I am not sure based on reading this part a couple of times...  You can absolutely, today, run the following command `choco install <path-to-packages.confing>` and Chocolatey CLI will install all of the packages that are defined in the file, there is no need to run any other commands.

> @csharpforevermore said...
Alternatively the choco export could support an option to export the actual choco CLI commands as well?

Given the ability to do the above, I am not sure that anything else is required for the export command in this area.  Although, now you have said this, it does make me think about exporting the currently installed packages into a specific Configuation Manager format, for example, into a Chef Recipe, or Ansible Playbook.

**Reply by csharpforevermore** [reply]
Hey there - thanks for your kind words.

I like how you can cite previous messages. (I couldn't figure out how to do it though sorry.)

So this idea mentioned above is that `choco export `and the website package script manager can read each other's output..  

For clarification, the Script Manager Tool (unsure about the exact name) is this useful tool highlighted on the left of the following screenshot:

![image](https://github.com/chocolatey/home/assets/4661508/e569def8-037e-4834-8e4e-0954cf3dd3a2)


So the following situations or user stories should help clarify things (forgive me for any ambiguity - I am on the spectrum and sometimes I have to repeat a concept a few times)

---

### Use case 1 

As a Chocolatey User, when I have a machine with Chocolatey installed on it, when I run the choco export command, I get a packages.config that I can then copy and paste or upload to the Chocolatey website .

### Use case 2

As a Chocolatey User, when I visit the website, I can upload a packages.config file and then edit it using the Script Manager Tool to amend  the list.  I can then download a new  packages.config file by clicking the "generate script" button. or something similar.

### Use case 3

As a Chocolatey User, when I am using an IaaS (Infrastructure as a Service) setup and I am manually setting up multiple environments (i.e. Staging, Pre-production/UAT, and Production) I can export my packages.config from my staging server, easily edit it using the website's Script Manager Tool to add or remove any unwanted applications, and then use this modified packages.config script to setup the next environment

---

The third use case is my own personal one  I figured there is a better way of editing the packages.config than using a text editor.  Also I like being able to use  the Script Manager Tool.

Lastly - is there any way I can contribute as a C# developer?  I love this product and have used it since the early days (I used to use the cinst command exclusively but the new way is good for standardisation)

Thanks again for reading!

**Reply by csharpforevermore** [reply]
Regarding the `choco export `changes, you expressed an enhanced version of what I was thinking. I thought of it producing a series of CLI commands instead of the XML packages.config.

So instead of outputting this:

`<?xml version="1.0" encoding="utf-8"?>
<packages>
   <package id="dotnetfx" />
</packages>`

.. the output would be (using a flag parameter to modify output perhaps?)...

`choco install dotnetfx`


Your idea of Chef and other installers, is a great further enhancement.

**Comment by pauby** [comment]
> I actually think that this is a good idea. @pauby what are your thoughts here? @st3phhays would this work?

I'm not sure if I really understand _why_ this is needed. As the `packages.config` is simply a text file, it can be edited with a standard text editor.

**Reply by csharpforevermore** [reply]
I am thinking just standardisation

Just noticed that I can click the "Generate Script" button and then "Save as .config File" so Use Case 2 is taken.

So the Script Manager Tool already outputs in different formats and I think `choco export `should do give the same output

Sure you can edit using a text editor but then why create GUIs in the first place?  Its the general statement about abstraction.  It's the same reason any application (including Chocolatey) has GUIs.

I find that I want to be able to search the Chocolatey website and add things to the current list.  Also I can put typos in the text editor but the GUI allows me to search for the exact name of the package and avoid poorly formatted XML.

Also why are you  guys using XML in this day and age of JSON?  Just like AS.NET Core moved from web.config to appsettings.json so too should this be a possible format of packages output maybe?  (i.e. packages.json )

**Reply by csharpforevermore** [reply]
I mean - what was the reason in creating the script package manager tool?  It's possible to just let people copy and paste the CLI right?

But it's a LOT easier to use the script package manager tool.

**Comment by pauby** [comment]
> Sure you can edit using a text editor but then why create GUIs in the first place? 

Script Builder was intended to allow you to _build_ a selection of packages. It was never intended as a GUI or editor of those packages.

> Also why are you guys using XML in this day and age of JSON?

`packages.config` is a defined XML format. We didn't create it. We extended it.

**Reply by csharpforevermore** [reply]
OK well that reply feels a little dismissive I must say - surprising given your quote about finding tech full of dismissiveness I must say - but thanks for reading and maybe someone will create it  In the meantime I think I'll start using the Chocolatey GUI rather than editing the packages.config using a text editor.  Maybe my workflow was unique or something.
